[Postgres] Use Incremental Snapshot Framework for Postgres CDC Connector

xiaom opened the pull request:
Co-Authored-By: Yaroslav Tkachenko <260702+sap1ens@users.noreply.github.com>

Hey @leonardBang,

This is our first PR resolving https://github.com/ververica/flink-cdc-connectors/issues/1163 😄 

The core functionality of DataStream is implemented under the package `com.ververica.cdc.connectors.postgres.source` with a similar layout to MySQL/Oracle's incremental snapshotting implementation.

```bash
source
├── PostgresChunkSplitter.java
├── PostgresConnectionPoolFactory.java
├── PostgresDialect.java
├── PostgresSourceBuilder.java
├── config
│   ├── PostgresSourceConfig.java
│   ├── PostgresSourceConfigFactory.java
│   └── PostgresSourceOptions.java
├── fetch
│   ├── PostgresScanFetchTask.java
│   ├── PostgresSourceFetchTaskContext.java
│   └── PostgresStreamFetchTask.java
├── offset
│   ├── PostgresOffset.java
│   └── PostgresOffsetFactory.java
└── utils
    ├── PgQueryUtils.java
    ├── PgSchema.java
    ├── PgTypeUtils.java
    └── TableDiscoveryUtils.java
```

The corresponding Table API can be enabled by setting `scan.incremental.snapshot.enabled=true`.

A few notes:

**package `io.debezium.connector.postgresql`**

The package is mostly getting around the limitation of using some Debezium classes directly. 

- `Utils.java`: a utility class to access some package-private methods of Debezium
- `PostgresObjectFactory.java`: a factory to create various Debezium object constructor which needs package private access
- ``PostgresConnection`.java`: copied from Debezium 1.6.4-final and modified to support injecting connection factory with Hikari connection pool.

**Major changes to CDC-base**

- JdbcSourceConfig: add a new field  `List<String> schemaList` and make it compatible with PostgreSQL and Debezium’s terminology.  (see Debezium’s [TableId](https://github.com/debezium/debezium/blob/v1.6.4.Final/debezium-core/src/main/java/io/debezium/relational/TableId.java)).
- DataSourceDialect: extend the CheckpointListener interface such that we can add a customized hook to commit offset
- SourceSplitSerializer: fixed a deserialization bug and check the `useCatalogBeforeSchema` flag (true by default)

**Other changes not related to PostgreSQL**

- Fix a bug where the high watermark is not set properly in various `*ScanFetchTask`

**Notes on Approximate Count query for Postgres**

We use the following query in chunk splitter to estimate the approximate count of rows

```sql
SELECT reltuples::bigint FROM pg_class WHERE oid = to_regclass('your_table_id')
```

The query requires a prior run of `VACCUM` or `ANALYZE` to get a good estimation.  For any PostgreSQL instances with [autovacuum](https://www.postgresql.org/docs/9.6/runtime-config-autovacuum.html) on, you won’t need to worry about it.

We are also actively working on supporting the `scan.newly-added-table.enabled` feature as @sap1ens [mentioned here](https://github.com/ververica/flink-cdc-connectors/issues/1152#issuecomment-1338500877).

Appreciate any feedback!

xiaom commented:
Update: 
- fix some issues of [e040533](https://github.com/ververica/flink-cdc-connectors/pull/1823/commits/e040533e15c3963f6c4b3c5ffb35553f8230e6dc) and set proper default implementation of several methods

leonardBang commented:
@xiaom We planed to open a contributor sync meeting to discuss the 2.4 roadmap, are you interested to join? please contact me if you'd like to.

xiaom commented:
Hey @leonardBang, thanks for the invitation! Yeah, I am interested. will DM you on Twitter.

1032851561 commented:
Does it support snapshot the new added tables?  I need this function, is it works good?

sap1ens commented:
> Does it support snapshot the new added tables? I need this function, is it works good?

There is a separate PR for that: https://github.com/ververica/flink-cdc-connectors/pull/1838

1032851561 commented:
> > Does it support snapshot the new added tables? I need this function, is it works good?
> 
> There is a separate PR for that: #1838

Nice, hope to merge as soon as possible.

ruanhang1993 reviewed (changes requested):
@xiaom , thanks for your work. I left some comments.

ruanhang1993 commented:
@xiaom Do you have time to rebase the master branch and make some updates? Please @me when you update this PR. Thanks ~

xiaom commented:
Hi @ruanhang1993, thanks for the review!  I will find some time to update the PR either later this week or next week.

ruanhang1993 commented:
Hi, @xiaom .

Is there any update about this PR?
We plan to release version 2.4.0 at June 14th. This feature is in this version. 
If any update is pushed, I will review again as soon as possible.

Thanks a lot~

xiaom commented:
Apologize for the delay in updating the PR. I've had some unexpected personal commitments come up.   
I'll do my best to get this PR updated.

xiaom commented on flink-connector-postgres-cdc/src/main/java/io/debezium/connector/postgresql/PostgresObjectFactory.java:
```diff
@@ -0,0 +1,135 @@
+                        StandardCharsets.UTF_8,
+                        config.getDecimalMode(),
+                        config.getTemporalPrecisionMode(),
+                        ZoneOffset.UTC,
```
I think it should be fine. Debezium's static factory method also use this fixed timezone.

https://github.com/debezium/debezium/blob/v1.6.4.Final/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java#L156

xiaom commented on flink-connector-postgres-cdc/src/main/java/com/ververica/cdc/connectors/postgres/source/fetch/PostgresSourceFetchTaskContext.java:
```diff
@@ -0,0 +1,277 @@
+    }
+
+    @Override
+    public RowType getSplitType(Table table) {
```
Good catch, changed to 

```
      Column splitColumn =
                PostgresChunkSplitter.getSplitColumn(table, sourceConfig.getChunkKeyColumn());
        return PostgresTypeUtils.getSplitType(splitColumn);
```

xiaom commented on flink-connector-postgres-cdc/src/main/java/com/ververica/cdc/connectors/postgres/source/fetch/PostgresStreamFetchTask.java:
```diff
@@ -0,0 +1,232 @@
+
+    private final StreamSplit split;
+    private volatile boolean taskRunning = false;
+    private volatile boolean stopped = false;
```
This came from https://github.com/ververica/flink-cdc-connectors/pull/1838 cc@sap1ens 

We want to prevent creating a replication slot when it already exists. To do so, we add a new method stop() to fetchTask.

xiaom commented on flink-connector-postgres-cdc/src/main/java/com/ververica/cdc/connectors/postgres/source/fetch/PostgresScanFetchTask.java:
```diff
@@ -0,0 +1,395 @@
+                        ctx.getReplicationConnection(),
+                        backfillSplit);
+        LOG.info("Execute backfillReadTask for split {}", split);
+        backfillReadTask.execute(new PostgresChangeEventSourceContext(), postgresOffsetContext);
```
Could you elaborate on what you mean here?

ruanhang1993 replied:
I see we could get the offsetContext from `ctx `. Why do we need to reload and use a new offsetContext. I do not get the point.

xiaom replied:
I see. This is similar to the [MySQL experimental connector](https://github.com/goldsky-io/goldsky-flink-cdc-connectors/blob/master/flink-cdc-base/src/test/java/com/ververica/cdc/connectors/base/experimental/fetch/MySqlScanFetchTask.java#L154-L158). 

We need to set the LSN to the low watermark when backfilling, otherwise it always tries to start with LSN 0/0.
In fact, we were using `ctx. getOffsetContext()` initially, and then found it causes issues:)

xiaom commented:
Hey @ruanhang1993, I've addressed some comments in [this commit](https://github.com/ververica/flink-cdc-connectors/pull/1823/commits/0abdc1e3e37746d3526610b40bf86a33ca6c9f58) for you to review. Let me know what you think. I have not rebased the branch yet. I will do it in the next update.

---

Also, I'd like to point out a caveat of this feature for any potential users: its scalability with large tables is not ideal. 

In the snapshotting phase, backfill tasks are created to capture new data changes. However, for larger tables, since snapshotting takes longer, WAL also grows larger and backfilling tasks will take significantly more time. 

Contrary to MySQL, where the process can be parallelized through additional binlog readers, this isn't straightforward for Postgres. To achieve similar parallelism, we would require more replication slots, a resource that is not advisable to overuse due to its limited availability.

In light of this, we implement a snapshot-only reader (with option `snapshot.mode=initial_only`) coupled with a stream reader (`snapshot.mode=never`) and some dedupe processes in our production environment. This approach allows us to parallelize snapshotting without increasing the number of replication slots. Just want to mention this in case anyone wants to use similar strategies.

ruanhang1993 commented on flink-connector-postgres-cdc/src/main/java/com/ververica/cdc/connectors/postgres/source/utils/TableDiscoveryUtils.java:
```diff
@@ -0,0 +1,58 @@
+            }
+        }
+
+        return capturedTables.stream().sorted().collect(Collectors.toCollection(ArrayList::new));
```
I think the order of tables may not be very useful. We could remove this sort.

xiaom replied:
Fixed:
```diff
-        return capturedTables.stream().sorted().collect(Collectors.toCollection(ArrayList::new));
+        return new ArrayList<>(capturedTables);
```

ruanhang1993 commented:
Hi, @xiaom . 
Thanks for the quick reply. I have replied the unclear comments.

About the problem you mentioned, the snapshot phase for the big table is actually a common pain point. 
Snapshot only + start from a specific binlog position is a good idea. But this way also have some limits.
- If there is a single parallelism in snapshot phase, we have to set an enough long alive time for binlogs.
- If there are multi parallelisms in snapshot phase, we can only provide the at least once semantic. The sink must support idempotent operations. And we also need to make sure the binlogs be alive.

The issue #1687 for mysql aims to the usage.

xiaom commented:
Hey @ruanhang1993,

I've rebased the PR.
---

Also, thanks for mentioning various solutions for parallelized snapshotting.  Good to know that this is a common pain point.

ruanhang1993 commented on docs/content/connectors/postgres-cdc.md:
```diff
@@ -183,12 +193,130 @@ Connector Options
+          <td>The max retry times that the connector should retry to build database server connection.</td>
+    </tr>
+    <tr>
+          <td>scan.incremental.snapshot.chunk.size</td>
```
Duplicate.

ruanhang1993 commented on flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/source/assigner/state/PendingSplitsStateSerializer.java:
```diff
@@ -358,6 +359,8 @@ private void writeTableIds(Collection<TableId> tableIds, DataOutputSerializer ou
         out.writeInt(size);
         for (TableId tableId : tableIds) {
+            boolean useCatalogBeforeSchema = SerializerUtils.shouldUseCatalogBeforeSchema(tableId);
+            out.writeBoolean(useCatalogBeforeSchema);
```
This will make the state in 2.3.0 not be able to be used in 2.4.0.
We should update the state serializer version and use a different logic.

ruanhang1993 commented on flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/utils/SerializerUtils.java:
```diff
@@ -77,4 +78,10 @@ public static Object serializedStringToObject(String serialized) {
         }
     }
+
+    public static boolean shouldUseCatalogBeforeSchema(TableId tableId) {
```
Add more comments for this method to introduce its usage.

ruanhang1993 commented on flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/source/reader/external/IncrementalSourceStreamFetcher.java:
```diff
@@ -97,25 +100,27 @@ public void submitTask(FetchTask<SourceSplitBase> fetchTask) {
+            sourceRecordsSet.add(new SourceRecords(sourceRecords));
+            return sourceRecordsSet.iterator();
+        } else {
+            return null;
```
Will this cause a NullPointerException?

ruanhang1993 commented on flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/source/reader/external/IncrementalSourceStreamFetcher.java:
```diff
@@ -85,6 +87,7 @@ public void submitTask(FetchTask<SourceSplitBase> fetchTask) {
                     try {
                         streamFetchTask.execute(taskContext);
```
Should we need to add `this.currentTaskRunning = false;` here ?

xiaom replied:
I think this is to ensure the task running is set properly after exception 

e.g., MySQL connector has a similar logic here
https://github.com/ververica/flink-cdc-connectors/blob/master/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/BinlogSplitReader.java#L130

ruanhang1993 commented on flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/source/reader/IncrementalSourceRecordEmitter.java:
```diff
@@ -107,15 +111,24 @@ protected void processElement(
+        }
+    }
+
+    private void updateStartingOffsetForSplit(SourceSplitState splitState, SourceRecord element) {
```
updateStreamSplitState ?

ruanhang1993 commented on flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/source/meta/offset/OffsetDeserializerSerializer.java:
```diff
@@ -82,7 +82,8 @@ default OffsetDeserializer createOffsetDeserializer() {
             final DataInputDeserializer in = new DataInputDeserializer(serialized);
-            TableId tableId = TableId.parse(in.readUTF());
+            boolean useCatalogBeforeSchema = in.readBoolean();
+            TableId tableId = TableId.parse(in.readUTF(), useCatalogBeforeSchema);
```
This change will make 2.3.0 state unusable. Need update to a new state version.

xiaom replied:
Could you give an example of how to do state version updates?

ruanhang1993 replied:
`FinishedSnapshotSplitInfo` is not a `SimpleVersionedSerializer`. We may need to change like this to allow 2.3.0 state.

FinishedSnapshotSplitInfo.java
```java
public byte[] serialize(final DataOutputSerializer out) throws IOException {
        out.writeUTF(this.getTableId().toString());
        out.writeUTF(this.getSplitId());
        out.writeUTF(SerializerUtils.rowToSerializedString(this.getSplitStart()));
        out.writeUTF(SerializerUtils.rowToSerializedString(this.getSplitEnd()));
        out.writeUTF(SerializerUtils.rowToSerializedString(this.offsetFactory));
        writeOffsetPosition(this.getHighWatermark(), out);
        boolean useCatalogBeforeSchema =
                SerializerUtils.shouldUseCatalogBeforeSchema(this.getTableId());
        out.writeBoolean(useCatalogBeforeSchema);
        return out.getCopyOfBuffer();
    }
```
OffsetDeserializerSerializer.java
```java
default FinishedSnapshotSplitInfo deserialize(byte[] serialized) {
        try {
            final DataInputDeserializer in = new DataInputDeserializer(serialized);
            String tableIdStr = in.readUTF();
            String splitId = in.readUTF();
            Object[] splitStart = serializedStringToRow(in.readUTF());
            Object[] splitEnd = serializedStringToRow(in.readUTF());
            OffsetFactory offsetFactory = (OffsetFactory) serializedStringToObject(in.readUTF());
            Offset highWatermark = readOffsetPosition(in);
            boolean useCatalogBeforeSchema = true;
            if (in.available() > 0) {
                useCatalogBeforeSchema = in.readBoolean();
            }
            TableId tableId = TableId.parse(tableIdStr, useCatalogBeforeSchema);
            in.releaseArrays();

            return new FinishedSnapshotSplitInfo(
                    tableId, splitId, splitStart, splitEnd, highWatermark, offsetFactory);
        } catch (IOException e) {
            throw new FlinkRuntimeException(e);
        }
    }
```

ruanhang1993 commented on flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/source/meta/offset/Offset.java:
```diff
@@ -39,7 +39,8 @@ public abstract class Offset implements Comparable<Offset>, Serializable {
 
     protected Map<String, String> offset;
```
If the return type is changed , this field should be changed too.

xiaom replied:
other connectors rely on the string value type/

e.g.,  if we change the filed, for oracle, it will have errors like this

> RedoLogOffset.java:57:26
java: incompatible types: capture#1 of ? cannot be converted to java.lang.String

ruanhang1993 replied:
How about changing like this.
Offset.java
```java
protected Map<String, ?> offset;

public Map<String, ?> getOffset() {
        return offset;
    }
```

Then all its subclass need to cast the value from this map in usage. 
For example : RedoLogOffset.java
```java
public String getScn() {
        return (String) offset.get(SCN_KEY);
    }

    public String getCommitScn() {
        return (String) offset.get(COMMIT_SCN_KEY);
    }

    public String getLcrPosition() {
        return (String) offset.get(LCR_POSITION_KEY);
    }
```

xiaom replied:
refactored to use String value type and only convert map when loading offset

ruanhang1993 reviewed (commented):
Hi, @xiaom . I have reviewed the cdc-base and will review pg cdc part later.
Would you mind take a look at the failed CI ? Thanks ~

ruanhang1993 commented on flink-connector-postgres-cdc/src/main/java/com/ververica/cdc/connectors/postgres/source/fetch/PostgresSourceFetchTaskContext.java:
```diff
@@ -0,0 +1,268 @@
+        return new TableId(null, schemaName, tableName);
+    }
+
+    public Offset getStreamOffset(SourceRecord sourceRecord) {
```
```
@Override
```

ruanhang1993 commented:
Hi, @xiaom .
I have finished reviewing the PR. Please take a look at it and make the CI succeed. Then we could merge this PR.
Thanks ~

xiaom commented:
I've addressed some review feedback ( emoji "👍" marked)  as part 1 [0861f46](https://github.com/ververica/flink-cdc-connectors/pull/1823/commits/0861f4607d2c65b34353b1225da7f43972a863c9). will continue the left one and fix CI later
